src/compose.ts

```diff
@@ -274,17 +285,96 @@ function normalizeArrayOfStrings(value: any[]): string[] {
-	if (colonIndex === -1) {
-		throw new ValidationError(`Invalid volume: '${serviceVolume}'`);
+function normalizeServiceBuild(
+	serviceBuild: string | BuildConfig,
+	networkNames: string[],
+): BuildConfig {
+	if (typeof serviceBuild === 'string') {
+		serviceBuild = { context: serviceBuild };
+	}
+	if (serviceBuild.args) {
+		serviceBuild.args = normalizeKeyValuePairs(serviceBuild.args);
+	}
+	if (serviceBuild.labels) {
+		serviceBuild.labels = normalizeKeyValuePairs(serviceBuild.labels);
+		validateLabels(serviceBuild.labels);
+	}
+	if (serviceBuild.extra_hosts && !_.isArray(serviceBuild.extra_hosts)) {
+		serviceBuild.extra_hosts = normalizeExtraHostObject(
+			serviceBuild.extra_hosts as any,
+		);
+	}
+	if (serviceBuild.isolation) {
+		throw new ValidationError('service.build.isolation is not allowed');
+	}
+	if (
+		serviceBuild.network &&
+		serviceBuild.network !== 'host' &&
+		serviceBuild.network !== 'none'
+	) {
+		if (networkNames.indexOf(serviceBuild.network) === -1) {
+			throw new ValidationError(
+				`Missing network definition for '${serviceBuild.network}'`,
+			);
+		}
 	}
-	const source = serviceVolume.slice(0, colonIndex);
-	if (path.parse(source).dir !== '') {
-		throw new ValidationError('Bind mounts are not allowed');
+	return serviceBuild;
+}
+
+function normalizeServiceVolumes(
+	serviceVolumes: any[],
+	volumeNames: string[],
+): string[] {
+	return serviceVolumes.map((volume) => {
+		const ref = normalizeServiceVolume(volume);
+		validateServiceVolume(ref, volumeNames);
+		return `${ref.source}:${ref.target}${ref.read_only ? ':ro' : ''}`;
+	});
+}
+
+function normalizeServiceVolume(serviceVolume: string | VolumeRef): VolumeRef {
+	let ref: VolumeRef = { type: 'volume', read_only: false };
+	if (typeof serviceVolume === 'string') {
+		const parts = serviceVolume.split(':');
+		if (parts.length < 2) {
+			throw new ValidationError(`Invalid volume: '${serviceVolume}'`);
+		}
+		ref.source = parts[0];
+		ref.target = parts[1];
+		if (path.parse(ref.source).dir) {
+			ref.type = 'bind';
+		}
+		if (parts[2] === 'ro') {
+			ref.read_only = true;
+		}
+	} else {
+		ref = serviceVolume;
 	}
-	if (volumeNames.indexOf(source) === -1) {
-		throw new ValidationError(`Missing volume definition for '${source}'`);
+	return ref;
+}
+
+function validateServiceVolume(
+	serviceVolume: VolumeRef,
+	volumeNames: string[],
+) {
+	switch (serviceVolume.type) {
+		case 'bind':
+			throw new ValidationError('Bind mounts are not allowed');
+		case 'tmpfs':
+			throw new ValidationError('Tmpfs mounts are not allowed');
+
+		case 'volume':
+			if (!serviceVolume.source) {
+				throw new ValidationError(`Missing volume source`);
+			}
+			if (volumeNames.indexOf(serviceVolume.source) === -1) {
+				throw new ValidationError(
+					`Missing volume definition for '${serviceVolume.source}'`,
+				);
+			}
+			if (serviceVolume.volume) {
+				throw new ValidationError(`Mount options are not allowed`);
```

Comment:
Nit:

```suggestion
				throw new ValidationError('Mount options are not allowed');
```

Same on line 368 above too.